subsys/net/lib/dhcpv6/dhcpv6.c

```diff
@@ -1327,6 +1380,49 @@ static int dhcpv6_find_status_code(struct net_pkt *pkt, uint16_t *status)
 	return ret;
 }
 
+static int dhcpv6_handle_dns_server_option(struct net_pkt *pkt)
+{
+	const struct sockaddr *dns_servers[MAX_DNS_SERVERS + 1] = { 0 };
+	struct sockaddr_in6 dnses[MAX_DNS_SERVERS] = { 0 };
+	uint16_t server_count = MAX_DNS_SERVERS;
+	struct dns_resolve_context *ctx;
+	struct net_pkt_cursor backup;
+	uint16_t length;
+	int ret, status;
+
+	net_pkt_cursor_backup(pkt, &backup);
+
+	ret = dhcpv6_find_option(pkt, DHCPV6_OPTION_CODE_OPTION_DNS_SERVERS,
+				 &length);
+	if (ret < 0) {
+		/* If the option is not present, don't report an error. */
+		ret = 0;
+		goto out;
+	}
+
+	ret = dhcpv6_parse_option_dns_servers(pkt, length, dnses, &server_count);
+	if (ret < 0) {
+		goto out;
+	}
+
+	for (uint8_t i = 0; i < server_count; i++) {
+		dnses[i].sin6_family = AF_INET6;
+		dns_servers[i] = (struct sockaddr *)&dnses[i];
+	}
+
+	ctx = dns_resolve_get_default();
+	status = dns_resolve_reconfigure(ctx, NULL, dns_servers);
+	if (status < 0) {
+		NET_DBG("Failed to reconfigure DNS resolver from DHCPv6 "
+			"option: %d", status);
```

Comment:
Should we return this error to caller?

Reply:
I wondered the same and decided not to. My reasoning is that even if we fail to register DNS servers internally for whatever reason, we could still finish processing the DHCPv6 message and use assigned IPv6 address.